### ISSUE TITLE
SELECT modifiers RENAME modifier

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#30411]
**Use case**

Make column names after huge amount of APPLY more predictable.

**Describe the solution you'd like**

```
WITH groupArray(time) AS time_arr
SELECT * EXCEPT time APPLY groupArray APPLY x -> arrayReverseSort((y, z) -> z, x, time_arr) APPLY arrayDifference APPLY arraySum 
FROM
(
    SELECT
        number AS time,
        number % 10 AS value,
        rand() % 10 AS value_2
    FROM numbers(100)
)

┌─arraySum(arrayDifference(arrayReverseSort(lambda(tuple(y, z), z), groupArray(value), time_arr)))─┬─arraySum(arrayDifference(arrayReverseSort(lambda(tuple(y, z), z), groupArray(value_2), time_arr)))─┐
│                                                                                               -9 │                                                                                                 -3 │
└──────────────────────────────────────────────────────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────┘


WITH groupArray(time) AS time_arr
SELECT * EXCEPT time APPLY groupArray APPLY x -> arrayReverseSort((y, z) -> z, x, time_arr) APPLY arrayDifference APPLY arraySum RENAME (col -> col || '_diff_sum')
FROM
(
    SELECT
        number AS time,
        number % 10 AS value,
        rand() % 10 AS value_2
    FROM numbers(100)
)

┌─value_diff_sum─┬─value_2_diff_sum─┐
│             -9 │               -3 │
└────────────────┴──────────────────┘
```

**Comment by amosbird** [comment]
@UnamedRus Good idea. We can make `RENAME` a terminal transformation and always refer to the root column names.

**Comment by amosbird** [comment]
> RENAME (col -> col || '_diff_sum')

It's complicated to support function evaluation inside parser. Let's instead use a similar syntax of the format function here.

There will be two implicit arguments, 1: column name, 2: column ordinal index.

`RENAME (col -> col || '_diff_sum')`:  `RENAME '{}_diff_sum'` or `RENAME '{0}_diff_sum'`

We can also numbering columns as we want: `RENAME 'diff_sum_{1}'`

**Comment by UnamedRus** [comment]
BTW, snowflake support rename transformer

https://github.com/ClickHouse/ClickHouse/pull/35660#issuecomment-1120194732

**Comment by yangguang760** [comment]
Hello,

I am writing to add my support and to respectfully urge the core team to prioritize the implementation of a column renaming feature.

The current behavior, where the result of an operation becomes a complex formula as the column name, creates significant friction in many practical use cases. This becomes particularly cumbersome when:

Applying multiple transformations in a query, as the column names become unreadable.

The results need to be referenced in subsequent parts of the same query (e.g., in a subquery or a JOIN), forcing the use of complex and error-prone aliases.

Outputting data to applications or users who expect clear, meaningful column names.

As mentioned in the thread, other modern data warehouses like Snowflake handle this seamlessly by allowing functions to be directly aliased, and their absence in ClickHouse is a notable gap in an otherwise excellent and high-performance database.

This issue has been open for nearly four years, and it's clear from the discussion that it is a pain point for multiple users in the community. The ability to rename columns inline is not just a syntactic sugar feature; it's a fundamental requirement for writing clean, maintainable, and complex SQL queries.

Could the team please provide an update on the potential timeline or any architectural considerations that have delayed this feature? Any movement on this would be greatly appreciated by myself and others following this issue.

Thank you for your hard work and consideration.

**Comment by niyue** [comment]
I opened a PR for this: https://github.com/ClickHouse/ClickHouse/pull/106033

It implements Snowflake-compatible explicit renaming syntax, such as `SELECT * RENAME value AS value_diff_sum`, and also supports the lambda-based batch renaming syntax discussed in this issue, such as `SELECT * RENAME (col -> concat(col, '_new'))`.

I also looked at DuckDB's pattern/template-based approach while implementing this. It seems less flexible than the lambda syntax for ClickHouse use cases, so I did not adopt it.

Feedback or suggestions on the PR would be very welcome.
